atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java

```diff
@@ -0,0 +1,255 @@
+        verifyStage();
+    }
+
+    @Test
+    public void getEventsForTransactionsReturnsSnapshotWithOldEvents() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processGetCommitTimestampsUpdate(COMMIT_UPDATE, SUCCESS);
+        eventCache.removeTransactionStateFromCache(START_TS);
+        verifyStage();
+
+        LockWatchStateUpdate success2 = LockWatchStateUpdate.success(LEADER, 10L, ImmutableList.of(LOCK_EVENT_2));
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS_2, success2);
+        verifyStage();
+
+        TransactionsLockWatchUpdate results = eventCache.getUpdateForTransactions(TIMESTAMPS_2, Optional.empty());
+        assertThat(results.clearCache()).isTrue();
+        assertThat(results.startTsToSequence()).containsExactlyInAnyOrderEntriesOf(
+                ImmutableMap.of(16L, IdentifiedVersion.of(LEADER, 10L)));
+        assertThat(results.events()).containsExactly(
+                LockWatchCreatedEvent.builder(ImmutableSet.of(REFERENCE),
+                        ImmutableSet.of(DESCRIPTOR, DESCRIPTOR_3)).build(6L),
+                LOCK_EVENT_2);
+    }
+
+    @Test
+    public void failedUpdateClearsAllCaches() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS_2, LockWatchStateUpdate.failed(LEADER));
+        verifyStage();
+    }
+
+    @Test
+    public void leaderChangeClearsCaches() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS_2,
+                LockWatchStateUpdate.success(EVENT2_UUID, 4L, ImmutableList.of()));
+        verifyStage();
+    }
+
+    @Test
+    public void removingEntriesRetentionsEventsInLog() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS_2, SUCCESS);
+        verifyStage();
+
+        eventCache.removeTransactionStateFromCache(START_TS);
+        verifyStage();
+    }
+
+    @Test
+    public void nonContiguousEventsThrows() {
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS, SNAPSHOT);
+        eventCache.processStartTransactionsUpdate(TIMESTAMPS_2,
+                LockWatchStateUpdate.success(LEADER, 10L, ImmutableList.of(WATCH_EVENT, LOCK_EVENT, LOCK_EVENT_2)));
+        eventCache.processStartTransactionsUpdate(ImmutableSet.of(23L, 55L),
+                LockWatchStateUpdate.success(LEADER, 20L,
+                        ImmutableList.of(UnlockEvent.builder(ImmutableSet.of()).build(20L))));
+        verifyStage();
+        assertThatThrownBy(() -> eventCache.removeTransactionStateFromCache(START_TS))
+                .isExactlyInstanceOf(SafeIllegalArgumentException.class)
+                .hasMessage("Events form a non-contiguous sequence");
+    }
+
+    @Test
+    public void snapshotMissedEventThrows() {
```

Comment:
Let's add a test where you race some updates, to simulate 2 concurrent calls (startTransactions and commitTimestamps): they will both start with the same version being sent to timelock, so second update should be ignored.

Reply:
Done